rust/kernel/kasync/executor.rs

```diff
@@ -83,18 +85,21 @@ pub fn ref_waker<T: 'static + ArcWake>(w: Arc<T>) -> Waker {
         )
     }
 
+    #[klint::preempt_count(expect = 0..)] // Required as `Waker::clone` must not sleep.
```

Comment:
Why `Waker::clone` must not sleep, could you elaborate?

Reply:
It is called by the callback passed to `init_waitqueue_func_entry`, which isn't allowed to sleep.

Reply:
Make senses, however, why do we need `clone` in `wake_callback`?

/me goes to search PR history

Reply:
Of course @wedsonaf may know the answer ;-)

Reply:
Looks like this is done to invoke the waker after unlocking the mutex protecting the waker is released. I would expect this is to prevent a deadlock or something like that.

Reply:
Looks to me that we can just do the following
```rust
        if let Some(mut guard) = s.waker.try_lock() {
            if let Some(w) = guard.take() {
                drop(guard);
                cloned.wake();
                w.wake();
                return 1;
            }
        |
```

Reply:
Yeah, that's what I suggested in the meeting. But I didn't make the change because I think the whole `NoWaitLock<Option<Waker>>` thing should be replaced with [`AtomicWaker`](https://docs.rs/futures/latest/futures/task/struct.AtomicWaker.html) instead.